### PR TITLE
Optimisation of Recrusive Fibonacci series

### PR DESCRIPTION
**Pull request by Manikant25**
# Recursive Fibonacci in C++ with Memoization

In an ordinary recursive function of the Fibonacci same function with the same parameter is called many times, so there are excessive recursive calls, but with memorization, we take an array of suitable size and initialize it with -1 and store the result of the unique recursive call in an array instead of -1.  So when we need to call a function again, we don't call we directly take its result from the array we have already stored and if does not exist then the function is called recursively.
There are no dependencies required as such, input is stored given in code.

Fixes: # (issue)

## Type of change
No change as of now.

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generates no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: snippets/c++/Recursive Fibonacci in C++ with Memoization/fibonacci.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
+# Recursive Fibonacci in C++ with Memoization
+
```
Comment (kaustubhgupta) [inline snippets/c++/Recursive Fibonacci in C++ with Memoization/fibonacci.md]: Can you describe this program in brief here? Maybe 2-3 lines?

File: snippets/c++/Recursive Fibonacci in C++ with Memoization/fibonacci.md [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+# Code
```
Comment (kaustubhgupta) [inline snippets/c++/Recursive Fibonacci in C++ with Memoization/fibonacci.md]: Make it level 2 heading (##)

File: snippets/c++/Recursive Fibonacci in C++ with Memoization/fibonacci.md [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+```C++
+#include<bits/stdc++.h>
+using namespace std;
+int F[10]; //Globally declared  or Static Array to store the value of called function
+           //It prevent from excessive recursion
+int mfib(int n)
+{
+ if(n<=1)
+ {
+ F[n]=n;
```
Comment (kaustubhgupta) [inline snippets/c++/Recursive Fibonacci in C++ with Memoization/fibonacci.md]: The whole code needs to be formatted properly, proper indentation

File: snippets/c++/Recursive Fibonacci in C++ with Memoization/fibonacci.md [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+ for(i=0;i<10;i++) //Initialising array F with -1
+ F[i]=-1;
+
+ printf("%d \n",mfib(5));
+ return 0;
+}
+```
+
+
+# Input
```
Comment (kaustubhgupta) [inline snippets/c++/Recursive Fibonacci in C++ with Memoization/fibonacci.md]: - Level 2 heading
- the input block is not ended

File: snippets/c++/Recursive Fibonacci in C++ with Memoization/fibonacci.md [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+```7```
+
+# Output
```
Comment (kaustubhgupta) [inline snippets/c++/Recursive Fibonacci in C++ with Memoization/fibonacci.md]: level 2

**Review by kaustubhgupta — CHANGES_REQUESTED** [review]

File: snippets/c++/Recursive Fibonacci in C++ with Memoization/fibonacci.md (hunk 1 above)
Comment (Manikant25) [reply snippets/c++/Recursive Fibonacci in C++ with Memoization/fibonacci.md]: I have described the program above already

Comment (kaustubhgupta) [reply snippets/c++/Recursive Fibonacci in C++ with Memoization/fibonacci.md]: I can't see in this file?

Comment (Manikant25) [reply snippets/c++/Recursive Fibonacci in C++ with Memoization/fibonacci.md]: okay I add

Comment (Manikant25) [reply snippets/c++/Recursive Fibonacci in C++ with Memoization/fibonacci.md]: Sir after making changes I will have to send pull request again and close this?

**Comment by Manikant25** [comment]
I have edited code , please review it

File: snippets/c++/Recursive Fibonacci in C++ with Memoization/fibonacci.md [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+  int i;
+  for (i = 0; i < 10; i++)
+    F[i] = -1;
+  printf("%d \n", mfib(7));
+  return 0;
+}
+```
+
+## Input
+```7```
```
Comment (kaustubhgupta) [inline snippets/c++/Recursive Fibonacci in C++ with Memoization/fibonacci.md]: I/O blocks are not formatted correctly. The number needs to be in the second line and in the third line the closing line, similar to how we define docstrings:
```python
"""
content here
"""
```

File: snippets/c++/Recursive Fibonacci in C++ with Memoization/fibonacci.md [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+## Output
+```13```
+
+## Contributed By
+|Name|Github Username|College Name|
+|---|---|---|
+|Manikant Rai|Manikant25|SRM Institute of Science And Technology|
+
```
Comment (kaustubhgupta) [inline snippets/c++/Recursive Fibonacci in C++ with Memoization/fibonacci.md]: remove this extra line

**Review by kaustubhgupta — CHANGES_REQUESTED** [review]

File: snippets/c++/Recursive Fibonacci in C++ with Memoization/fibonacci.md (hunk 6 above)
Comment (sakshamarora1) [inline snippets/c++/Recursive Fibonacci in C++ with Memoization/fibonacci.md]:
~~~suggestion
```
7
```
~~~

File: snippets/c++/Recursive Fibonacci in C++ with Memoization/fibonacci.md [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+  printf("%d \n", mfib(7));
+  return 0;
+}
+```
+
+## Input
+```7```
+
+## Output
+```13```
```
Comment (sakshamarora1) [inline snippets/c++/Recursive Fibonacci in C++ with Memoization/fibonacci.md]:
~~~suggestion
```
13
```
~~~

**Review by sakshamarora1 — CHANGES_REQUESTED** [review]
Looks great! Just these few changes along with what kaustubh said! :D

**Comment by Manikant25** [comment]
Thank you for reviewing, I have made changes. Have a look.

**Review by kaustubhgupta — APPROVED** [review]
LGTM now

File: snippets/c++/Recursive Fibonacci in C++ with Memoization/fibonacci.md [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+
+    return F[n];
+  }
+}
+
+int main() {
+  int i;
+  for (i = 0; i < 10; i++)
+    F[i] = -1;
+  printf("%d \n", mfib(7));
```
Comment (sakshamarora1) [inline snippets/c++/Recursive Fibonacci in C++ with Memoization/fibonacci.md]: So, I noticed that you aren't taking any input here. Can you update the program to take an input with a prompt here and also reflect that in the input block below?

File: snippets/c++/Recursive Fibonacci in C++ with Memoization/fibonacci.md [hunk 10]
```diff
@@ -0,0 +1,49 @@
+# Recursive Fibonacci in C++ with Memoization
+
+## Description
+In an ordinary recursive function of the Fibonacci same function with the same parameter is called many times, so there are excessive recursive calls, but with memorization, we take an array of suitable size and initialize it with -1 and store the result of the unique recursive call in an array instead of -1. So when we need to call a function again, we don't call we directly take its result from the array we have already stored and if does not exist then the function is called recursively.
```
Comment (sakshamarora1) [inline snippets/c++/Recursive Fibonacci in C++ with Memoization/fibonacci.md]:
```suggestion
In an ordinary recursive function of the Fibonacci same function with the same parameter is called many times, so there are excessive recursive calls, but with memoization, an array of suitable size is taken and each value is initialized with -1. The result of the unique recursive call in an array is then stored replacing -1. So, when the function needs to be called again, it isn't called, instead, the result is directly taken from the array already stored and if it does not exists then the function is called recursively.
```

Comment (sakshamarora1) [reply snippets/c++/Recursive Fibonacci in C++ with Memoization/fibonacci.md]: Also add it in the description that this code works for the first 100 numbers in the fibonacci sequence.

File: snippets/c++/Recursive Fibonacci in C++ with Memoization/fibonacci.md [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+
+## Description
+In an ordinary recursive function of the Fibonacci same function with the same parameter is called many times, so there are excessive recursive calls, but with memorization, we take an array of suitable size and initialize it with -1 and store the result of the unique recursive call in an array instead of -1. So when we need to call a function again, we don't call we directly take its result from the array we have already stored and if does not exist then the function is called recursively.  
+
+## Code
+```C++
+#include<bits/stdc++.h>
+
+using namespace std;
+int F[10]; //Globally declared  or Static Array to store the value of called function
```
Comment (sakshamarora1) [inline snippets/c++/Recursive Fibonacci in C++ with Memoization/fibonacci.md]: Change this to extend to 100 numbers in the sequence

File: snippets/c++/Recursive Fibonacci in C++ with Memoization/fibonacci.md [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+      F[n - 1] = mfib(n - 1);
+    F[n] = F[n - 2] + F[n - 1];
+
+    return F[n];
+  }
+}
+
+int main() {
+  int i;
+  for (i = 0; i < 10; i++)
```
Comment (sakshamarora1) [inline snippets/c++/Recursive Fibonacci in C++ with Memoization/fibonacci.md]:
```suggestion
  for (i = 0; i < 100; i++)
```

File: snippets/c++/Recursive Fibonacci in C++ with Memoization/fibonacci.md [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+## Input
+```
+7
+```
+
+## Output
+```
+13
+```
+## Contributed By
```
Comment (sakshamarora1) [inline snippets/c++/Recursive Fibonacci in C++ with Memoization/fibonacci.md]:
~~~suggestion
```

## Contributed By
~~~

**Review by sakshamarora1 — CHANGES_REQUESTED** [review]
The file doesn't have any output image/gif so we can keep the file in `c++` itself instead of having an extra folder level. Also rename it to `RecursiveFibonacciWithMemoization.md`

File: snippets/c++/Recursive Fibonacci in C++ with Memoization/fibonacci.md (hunk 13 above)
Comment (Manikant25) [reply snippets/c++/Recursive Fibonacci in C++ with Memoization/fibonacci.md]: Done

File: snippets/c++/RecursiveFibonacciWithMemoization.md [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+
+    return F[n];
+  }
+}
+
+int main() {
+  int i,x;
+  for (i = 0; i < 100; i++)
+    F[i] = -1;
+  cout << "Enter the number of terms of series : ";
```
Comment (sakshamarora1) [inline snippets/c++/RecursiveFibonacciWithMemoization.md]:
```suggestion
  cout << "Enter the index(max:100) : ";  
```

File: snippets/c++/RecursiveFibonacciWithMemoization.md [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+  cout << "Enter the number of terms of series : ";  
+  cin>>x;
+  cout <<mfib(x);
+  return 0;
+}
+```
+
+## Input
+```
+Enter the number of terms of series : 7
```
Comment (sakshamarora1) [inline snippets/c++/RecursiveFibonacciWithMemoization.md]:
```suggestion
Enter the index(max:100) : 7
```

File: snippets/c++/RecursiveFibonacciWithMemoization.md [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+  }
+}
+
+int main() {
+  int i,x;
+  for (i = 0; i < 100; i++)
+    F[i] = -1;
+  cout << "Enter the number of terms of series : ";  
+  cin>>x;
+  cout <<mfib(x);
```
Comment (sakshamarora1) [inline snippets/c++/RecursiveFibonacciWithMemoization.md]:
```suggestion
  cout << "The " << x << "th term in the fibonacci sequence is : " << mfib(x);
```

File: snippets/c++/RecursiveFibonacciWithMemoization.md [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+}
+```
+
+## Input
+```
+Enter the number of terms of series : 7
+```
+## Output
+```
+13
```
Comment (sakshamarora1) [inline snippets/c++/RecursiveFibonacciWithMemoization.md]:
```suggestion
The 7th term in the fibonacci sequence is : 13
```

Comment (Manikant25) [reply snippets/c++/RecursiveFibonacciWithMemoization.md]: Now is it fine?

Comment (sakshamarora1) [reply snippets/c++/RecursiveFibonacciWithMemoization.md]: Yes! :partying_face:

**Pull request merged**